File: strings/anagram.cpp
#include <iostream>
#include <string>


std::string anagram1(std::string s1, std::string s2){
    
    std::string positive = "Anagram";
    std::string negative = "Not Anagram";
 
    if(s1.length() != s2.length())
        return negative;
    
    if(s1.length() < 2 && s1 != s2)    
      return negative;

    std::sort(s1.begin(),s1.end());
    std::sort(s2.begin(),s2.end());

    if(s1 != s2)
        return negative;
    
    return positive;
}

std::string anagram2(std::string s1, std::string s2) {

    std::string positive = "Anagram";
    std::string negative = "Not Anagram";

    if(s1.length() != s2.length())
        return negative;
    
   if(s1.length() < 2 && s1 != s2)    
      return negative;


    int letters[256];    
    
    for(int i = 0; i < 256; i++) {
         letters[i] = 0;
    }

    for(int i = 0; i < s1.length(); i++) {
        letters[s1[i]]++;
        letters[s2[i]]--;
    }

    for(int i = 0; i < 256; i++) {
        if(letters[i] != 0) {
            return negative;
        }
    }
    return positive;
}

int main () {
    std::string s1 = "iceman";
    std::string s2 = "cinema";
    
    std::cout << "Method 1: "  << anagram1(s1,s2) << std::endl;
    std::cout << "Method 2: " << anagram2(s1,s2) << std::endl;
    
    return 0;
}

File: strings/anagram.java

class anagram 
{
   public static void main(String args[])
   {
      String a = "cinema";
      String b = "iceman";
      anagram Obj = new anagram();
      System.out.println(Obj.anagram1(a,b));
      System.out.println(Obj.anagram2(a,b));
   }

   public String sort(String s)
   {
      char[] array = s.toCharArray();
      java.util.Arrays.sort(array);
      return new String(array);
   }  
   public boolean anagram1(String a, String b)
   {
      if(a.length() != b.length())
         return false;

      return sort(a).equals(sort(b));

   }

   public boolean anagram2(String a,String b) 
   {
      int[] array = new int[256];

      int len = a.length();
      if(len != b.length())
         return false;
      
      for(int i = 0; i < len; i++)
      {
         array[a.charAt(i)]++;
         array[b.charAt(i)]--;
      }

      for(int i = 0; i < 256; i++)
         if(array[i] != 0)
            return false;
      return true;
   }

}
